// @curvefi/api/lib/curve.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.curve = void 0;
var ethers_1 = require("ethers");
var ethcall_1 = require("ethcall");
var factory_1 = require("./factory/factory");
var factory_api_1 = require("./factory/factory-api");
var factory_crypto_1 = require("./factory/factory-crypto");
var ERC20_json_1 = __importDefault(require("./constants/abis/ERC20.json"));
var cERC20_json_1 = __importDefault(require("./constants/abis/cERC20.json"));
var yERC20_json_1 = __importDefault(require("./constants/abis/yERC20.json"));
var minter_json_1 = __importDefault(require("./constants/abis/minter.json"));
var votingescrow_json_1 = __importDefault(require("./constants/abis/votingescrow.json"));
var address_provider_json_1 = __importDefault(require("./constants/abis/address_provider.json"));
var gaugecontroller_json_1 = __importDefault(require("./constants/abis/gaugecontroller.json"));
var router_json_1 = __importDefault(require("./constants/abis/router.json"));
var deposit_and_stake_json_1 = __importDefault(require("./constants/abis/deposit_and_stake.json"));
var registry_exchange_json_1 = __importDefault(require("./constants/abis/registry_exchange.json"));
var streamer_json_1 = __importDefault(require("./constants/abis/streamer.json"));
var factory_json_1 = __importDefault(require("./constants/abis/factory.json"));
var factory_crypto_json_1 = __importDefault(require("./constants/abis/factory-crypto.json"));
var pools_1 = require("./constants/pools");
var ethereum_1 = require("./constants/coins/ethereum");
var polygon_1 = require("./constants/coins/polygon");
var aliases_1 = require("./constants/aliases");
var utils_1 = require("./constants/utils");
var CONSTANTS = {
    1: {
        ALIASES: aliases_1.ALIASES_ETHEREUM,
        POOLS_DATA: pools_1.POOLS_DATA_ETHEREUM,
        COINS: ethereum_1.COINS_ETHEREUM,
        cTokens: ethereum_1.cTokensEthereum,
        yTokens: ethereum_1.yTokensEthereum,
        ycTokens: ethereum_1.ycTokensEthereum,
        aTokens: ethereum_1.aTokensEthereum,
    },
    137: {
        ALIASES: aliases_1.ALIASES_POLYGON,
        POOLS_DATA: pools_1.POOLS_DATA_POLYGON,
        COINS: polygon_1.COINS_POLYGON,
        cTokens: polygon_1.cTokensPolygon,
        yTokens: polygon_1.yTokensPolygon,
        ycTokens: polygon_1.ycTokensPolygon,
        aTokens: polygon_1.aTokensPolygon,
    },
};
var Curve = /** @class */ (function () {
    function Curve() {
        // @ts-ignore
        this.provider = null;
        // @ts-ignore
        this.signer = null;
        this.signerAddress = '';
        this.chainId = 0;
        // @ts-ignore
        this.multicallProvider = null;
        this.contracts = {};
        this.feeData = {};
        this.constantOptions = { gasLimit: 12000000 };
        this.options = {};
        this.constants = {
            ALIASES: {},
            POOLS_DATA: {},
            FACTORY_POOLS_DATA: {},
            CRYPTO_FACTORY_POOLS_DATA: {},
            COINS: {},
            DECIMALS: {},
            GAUGES: [],
        };
    }
    Curve.prototype.init = function (providerType, providerSettings, options // gasPrice in Gwei
    ) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var network, _a, _b, cTokens, yTokens, ycTokens, aTokens, customAbiTokens, _c, _i, _d, pool, _e, _f, coinAddr, _g, _h, coinAddr, _j, _k, coinAddr, _l, _m, rewardTokenAddr, addressProviderContract, _o;
            return __generator(this, function (_p) {
                switch (_p.label) {
                    case 0:
                        // @ts-ignore
                        this.provider = null;
                        // @ts-ignore
                        this.signer = null;
                        this.signerAddress = '';
                        this.chainId = 0;
                        // @ts-ignore
                        this.multicallProvider = null;
                        this.contracts = {};
                        this.feeData = {};
                        this.constantOptions = { gasLimit: 12000000 };
                        this.options = {};
                        this.constants = {
                            ALIASES: {},
                            POOLS_DATA: {},
                            FACTORY_POOLS_DATA: {},
                            CRYPTO_FACTORY_POOLS_DATA: {},
                            COINS: {},
                            DECIMALS: {},
                            GAUGES: [],
                        };
                        // JsonRpc provider
                        if (providerType.toLowerCase() === 'JsonRpc'.toLowerCase()) {
                            providerSettings = providerSettings;
                            if (providerSettings.url) {
                                this.provider = this.provider = new ethers_1.ethers.providers.JsonRpcProvider(providerSettings.url);
                            }
                            else {
                                this.provider = new ethers_1.ethers.providers.JsonRpcProvider('http://localhost:8545/');
                            }
                            if (providerSettings.privateKey) {
                                this.signer = new ethers_1.ethers.Wallet(providerSettings.privateKey, this.provider);
                            }
                            else {
                                this.signer = this.provider.getSigner();
                            }
                            // Web3 provider
                        }
                        else if (providerType.toLowerCase() === 'Web3'.toLowerCase()) {
                            providerSettings = providerSettings;
                            this.provider = new ethers_1.ethers.providers.Web3Provider(providerSettings.externalProvider);
                            this.signer = this.provider.getSigner();
                            // Infura provider
                        }
                        else if (providerType.toLowerCase() === 'Infura'.toLowerCase()) {
                            providerSettings = providerSettings;
                            this.provider = new ethers_1.ethers.providers.InfuraProvider(providerSettings.network, providerSettings.apiKey);
                            this.signer = null;
                            // Alchemy provider
                        }
                        else if (providerType.toLowerCase() === 'Alchemy'.toLowerCase()) {
                            providerSettings = providerSettings;
                            this.provider = new ethers_1.ethers.providers.AlchemyProvider(providerSettings.network, providerSettings.apiKey);
                            this.signer = null;
                        }
                        else {
                            throw Error('Wrong providerType');
                        }
                        _a = this.provider.network;
                        if (_a) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.provider._networkPromise];
                    case 1:
                        _a = (_p.sent());
                        _p.label = 2;
                    case 2:
                        network = _a;
                        console.log("CURVE-JS IS CONNECTED TO NETWORK:", network);
                        this.chainId = network.chainId === 1337 ? 1 : network.chainId;
                        this.constants.ALIASES = CONSTANTS[this.chainId].ALIASES;
                        this.constants.POOLS_DATA = CONSTANTS[this.chainId].POOLS_DATA;
                        this.constants.COINS = CONSTANTS[this.chainId].COINS;
                        this.constants.DECIMALS = (0, utils_1.extractDecimals)(this.constants.POOLS_DATA);
                        this.constants.GAUGES = (0, utils_1.extractGauges)(this.constants.POOLS_DATA);
                        _b = [
                            CONSTANTS[this.chainId].cTokens,
                            CONSTANTS[this.chainId].yTokens,
                            CONSTANTS[this.chainId].ycTokens,
                            CONSTANTS[this.chainId].aTokens,
                        ], cTokens = _b[0], yTokens = _b[1], ycTokens = _b[2], aTokens = _b[3];
                        customAbiTokens = __spreadArray(__spreadArray(__spreadArray(__spreadArray([], cTokens, true), yTokens, true), ycTokens, true), aTokens, true);
                        this.multicallProvider = new ethcall_1.Provider();
                        return [4 /*yield*/, this.multicallProvider.init(this.provider)];
                    case 3:
                        _p.sent();
                        if (!this.signer) return [3 /*break*/, 5];
                        _c = this;
                        return [4 /*yield*/, this.signer.getAddress()];
                    case 4:
                        _c.signerAddress = _p.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        this.signerAddress = '';
                        _p.label = 6;
                    case 6:
                        this.feeData = { gasPrice: options.gasPrice, maxFeePerGas: options.maxFeePerGas, maxPriorityFeePerGas: options.maxPriorityFeePerGas };
                        return [4 /*yield*/, this.updateFeeData()];
                    case 7:
                        _p.sent();
                        for (_i = 0, _d = Object.values(this.constants.POOLS_DATA); _i < _d.length; _i++) {
                            pool = _d[_i];
                            this.contracts[pool.swap_address] = {
                                contract: new ethers_1.Contract(pool.swap_address, pool.swap_abi, this.signer || this.provider),
                                multicallContract: new ethcall_1.Contract(pool.swap_address, pool.swap_abi),
                            };
                            if (pool.token_address !== pool.swap_address) {
                                this.contracts[pool.token_address] = {
                                    contract: new ethers_1.Contract(pool.token_address, ERC20_json_1.default, this.signer || this.provider),
                                    multicallContract: new ethcall_1.Contract(pool.token_address, ERC20_json_1.default),
                                };
                            }
                            this.contracts[pool.gauge_address] = {
                                contract: new ethers_1.Contract(pool.gauge_address, pool.gauge_abi, this.signer || this.provider),
                                multicallContract: new ethcall_1.Contract(pool.gauge_address, pool.gauge_abi),
                            };
                            if (pool.deposit_address && !this.contracts[pool.deposit_address]) {
                                this.contracts[pool.deposit_address] = {
                                    contract: new ethers_1.Contract(pool.deposit_address, pool.deposit_abi, this.signer || this.provider),
                                    multicallContract: new ethcall_1.Contract(pool.deposit_address, pool.deposit_abi),
                                };
                            }
                            for (_e = 0, _f = pool.underlying_coin_addresses; _e < _f.length; _e++) {
                                coinAddr = _f[_e];
                                this.contracts[coinAddr] = {
                                    contract: new ethers_1.Contract(coinAddr, ERC20_json_1.default, this.signer || this.provider),
                                    multicallContract: new ethcall_1.Contract(coinAddr, ERC20_json_1.default),
                                };
                            }
                            for (_g = 0, _h = pool.wrapped_coin_addresses; _g < _h.length; _g++) {
                                coinAddr = _h[_g];
                                if (customAbiTokens.includes(coinAddr))
                                    continue;
                                this.contracts[coinAddr] = {
                                    contract: new ethers_1.Contract(coinAddr, ERC20_json_1.default, this.signer || this.provider),
                                    multicallContract: new ethcall_1.Contract(coinAddr, ERC20_json_1.default),
                                };
                            }
                            // TODO add all coins
                            for (_j = 0, _k = pool.wrapped_coin_addresses; _j < _k.length; _j++) {
                                coinAddr = _k[_j];
                                if (cTokens.includes(coinAddr)) {
                                    this.contracts[coinAddr] = {
                                        contract: new ethers_1.Contract(coinAddr, cERC20_json_1.default, this.signer || this.provider),
                                        multicallContract: new ethcall_1.Contract(coinAddr, cERC20_json_1.default),
                                    };
                                }
                                if (aTokens.includes(coinAddr)) {
                                    this.contracts[coinAddr] = {
                                        contract: new ethers_1.Contract(coinAddr, ERC20_json_1.default, this.signer || this.provider),
                                        multicallContract: new ethcall_1.Contract(coinAddr, ERC20_json_1.default),
                                    };
                                }
                                if (yTokens.includes(coinAddr) || ycTokens.includes(coinAddr)) {
                                    this.contracts[coinAddr] = {
                                        contract: new ethers_1.Contract(coinAddr, yERC20_json_1.default, this.signer || this.provider),
                                        multicallContract: new ethcall_1.Contract(coinAddr, yERC20_json_1.default),
                                    };
                                }
                            }
                            if (pool.reward_contract) {
                                this.contracts[pool.reward_contract] = {
                                    contract: new ethers_1.Contract(pool.reward_contract, streamer_json_1.default, this.signer || this.provider),
                                    multicallContract: new ethcall_1.Contract(pool.reward_contract, streamer_json_1.default),
                                };
                            }
                            for (_l = 0, _m = pool.reward_tokens || []; _l < _m.length; _l++) {
                                rewardTokenAddr = _m[_l];
                                this.contracts[rewardTokenAddr] = {
                                    contract: new ethers_1.Contract(rewardTokenAddr, ERC20_json_1.default, this.signer || this.provider),
                                    multicallContract: new ethcall_1.Contract(rewardTokenAddr, ERC20_json_1.default),
                                };
                            }
                        }
                        this.contracts[this.constants.ALIASES.crv] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.crv, ERC20_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.crv, ERC20_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.crv.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.crv, ERC20_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.crv, ERC20_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.minter] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.minter, minter_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.minter, minter_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.minter.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.minter, minter_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.minter, minter_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.voting_escrow] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.voting_escrow, votingescrow_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.voting_escrow, votingescrow_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.voting_escrow.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.voting_escrow, votingescrow_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.voting_escrow, votingescrow_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.address_provider] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.address_provider, address_provider_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.address_provider, address_provider_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.address_provider.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.address_provider, address_provider_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.address_provider, address_provider_json_1.default),
                        };
                        addressProviderContract = this.contracts[this.constants.ALIASES.address_provider].contract;
                        _o = this.constants.ALIASES;
                        return [4 /*yield*/, addressProviderContract.get_address(2, this.constantOptions)];
                    case 8:
                        _o.registry_exchange = _p.sent();
                        this.contracts[this.constants.ALIASES.registry_exchange] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.registry_exchange, registry_exchange_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.registry_exchange, registry_exchange_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.registry_exchange.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.registry_exchange, registry_exchange_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.registry_exchange, registry_exchange_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.gauge_controller] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.gauge_controller, gaugecontroller_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.gauge_controller, gaugecontroller_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.gauge_controller.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.gauge_controller, gaugecontroller_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.gauge_controller, gaugecontroller_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.router] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.router, router_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.router, router_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.router.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.router, router_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.router, router_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.deposit_and_stake] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.deposit_and_stake, deposit_and_stake_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.deposit_and_stake, deposit_and_stake_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.deposit_and_stake.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.deposit_and_stake, deposit_and_stake_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.deposit_and_stake, deposit_and_stake_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.factory] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.factory, factory_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.factory, factory_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.factory.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.factory, factory_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.factory, factory_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.crypto_factory] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.crypto_factory, factory_crypto_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.crypto_factory, factory_crypto_json_1.default),
                        };
                        this.contracts[this.constants.ALIASES.crypto_factory.toLowerCase()] = {
                            contract: new ethers_1.Contract(this.constants.ALIASES.crypto_factory, factory_crypto_json_1.default, this.signer || this.provider),
                            multicallContract: new ethcall_1.Contract(this.constants.ALIASES.crypto_factory, factory_crypto_json_1.default),
                        };
                        return [2 /*return*/];
                }
            });
        });
    };
    Curve.prototype.fetchFactoryPools = function (useApi) {
        if (useApi === void 0) { useApi = true; }
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (!useApi) return [3 /*break*/, 2];
                        _a = this.constants;
                        _b = utils_1.lowerCasePoolDataAddresses;
                        return [4 /*yield*/, factory_api_1.getFactoryPoolsDataFromApi.call(this, false)];
                    case 1:
                        _a.FACTORY_POOLS_DATA = _b.apply(void 0, [_e.sent()]);
                        return [3 /*break*/, 4];
                    case 2:
                        _c = this.constants;
                        _d = utils_1.lowerCasePoolDataAddresses;
                        return [4 /*yield*/, factory_1.getFactoryPoolData.call(this)];
                    case 3:
                        _c.FACTORY_POOLS_DATA = _d.apply(void 0, [_e.sent()]);
                        _e.label = 4;
                    case 4:
                        this.constants.DECIMALS = __assign(__assign({}, this.constants.DECIMALS), (0, utils_1.extractDecimals)(this.constants.FACTORY_POOLS_DATA));
                        this.constants.GAUGES = __spreadArray(__spreadArray([], this.constants.GAUGES, true), (0, utils_1.extractGauges)(this.constants.FACTORY_POOLS_DATA), true);
                        return [2 /*return*/];
                }
            });
        });
    };
    Curve.prototype.fetchCryptoFactoryPools = function (useApi) {
        if (useApi === void 0) { useApi = true; }
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (this.chainId !== 1 && this.chainId !== 1337)
                            return [2 /*return*/];
                        if (!useApi) return [3 /*break*/, 2];
                        _a = this.constants;
                        _b = utils_1.lowerCasePoolDataAddresses;
                        return [4 /*yield*/, factory_api_1.getFactoryPoolsDataFromApi.call(this, true)];
                    case 1:
                        _a.CRYPTO_FACTORY_POOLS_DATA = _b.apply(void 0, [_e.sent()]);
                        return [3 /*break*/, 4];
                    case 2:
                        _c = this.constants;
                        _d = utils_1.lowerCasePoolDataAddresses;
                        return [4 /*yield*/, factory_crypto_1.getCryptoFactoryPoolData.call(this)];
                    case 3:
                        _c.CRYPTO_FACTORY_POOLS_DATA = _d.apply(void 0, [_e.sent()]);
                        _e.label = 4;
                    case 4:
                        this.constants.DECIMALS = __assign(__assign({}, this.constants.DECIMALS), (0, utils_1.extractDecimals)(this.constants.CRYPTO_FACTORY_POOLS_DATA));
                        this.constants.GAUGES = __spreadArray(__spreadArray([], this.constants.GAUGES, true), (0, utils_1.extractGauges)(this.constants.CRYPTO_FACTORY_POOLS_DATA), true);
                        return [2 /*return*/];
                }
            });
        });
    };
    Curve.prototype.setCustomFeeData = function (customFeeData) {
        this.feeData = __assign(__assign({}, this.feeData), customFeeData);
    };
    Curve.prototype.updateFeeData = function () {
        return __awaiter(this, void 0, void 0, function () {
            var feeData, _a, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.provider.getFeeData()];
                    case 1:
                        feeData = _d.sent();
                        if (!(feeData.maxFeePerGas === null || feeData.maxPriorityFeePerGas === null)) return [3 /*break*/, 6];
                        delete this.options.maxFeePerGas;
                        delete this.options.maxPriorityFeePerGas;
                        _a = this.options;
                        if (!(this.feeData.gasPrice !== undefined)) return [3 /*break*/, 2];
                        _b = ethers_1.ethers.utils.parseUnits(this.feeData.gasPrice.toString(), "gwei");
                        return [3 /*break*/, 5];
                    case 2:
                        _c = feeData.gasPrice;
                        if (_c) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.provider.getGasPrice()];
                    case 3:
                        _c = (_d.sent());
                        _d.label = 4;
                    case 4:
                        _b = (_c);
                        _d.label = 5;
                    case 5:
                        _a.gasPrice = _b;
                        return [3 /*break*/, 7];
                    case 6:
                        delete this.options.gasPrice;
                        this.options.maxFeePerGas = this.feeData.maxFeePerGas !== undefined ?
                            ethers_1.ethers.utils.parseUnits(this.feeData.maxFeePerGas.toString(), "gwei") :
                            feeData.maxFeePerGas;
                        this.options.maxPriorityFeePerGas = this.feeData.maxPriorityFeePerGas !== undefined ?
                            ethers_1.ethers.utils.parseUnits(this.feeData.maxPriorityFeePerGas.toString(), "gwei") :
                            feeData.maxPriorityFeePerGas;
                        _d.label = 7;
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    return Curve;
}());
exports.curve = new Curve();
